module-2/lab-probability-distribution/your_code/main.ipynb

```diff
@@ -70,31 +70,31 @@
       "\n",
```

Comment:
- Sometimes it's useful to seed the random number generator. Seeding a pseudo-random number generator gives it its first "previous" value. Each seed value will correspond to a sequence of generated values for a given random number generator. That is, if you provide the same seed twice, you get the same sequence of numbers twice. https://machinelearningmastery.com/how-to-generate-random-numbers-in-python/
- When you say "All depends on the threshold value we define for our hipothesis to check wherever we do not have a normal", please use confidence level instead of threshold. It's just notation but it's important in statistics. 
- Normality tests are quite sensitive to sample size thus for a sample size below a couple of thousands it should be a better idea to evaluate normality through a plot, in particular a quantile-quantile plot (QQ plot). A Q-Q plot is a scatterplot created by plotting two sets of quantiles against one another. It plots the quantiles of our sample vs the theoretical quantiles of the distribution type we want to test (this case a normal distribution) How to do this in Python? https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.probplot.html.
-There are 2 other normality tests (additional to D'Agostino k^2 test, normaltest in Python) that are very used in to test normality of our data. These are Kolmogorov-Smirnof and Shapiro-Wilk tests. I encourage you to deepen on these type of tests. In case you are interested, I can provide you some useful resources.
- Regarding this : "All distributions have the mean on 50 or near it. This fits with central theorem where taking a random sample of elements of a distribution its mean values are near the mean, and when we take more and more samples it returns on getting closer to the defined mean". 
- Good one!! Thanks for setting x,y labels and title in plots. 
```Python
nsubs=len(rand_samp)
fig,axes=plt.subplots(1,nsubs,figsize=(25,6))
counter=-1
for distrib,samp in rand_samp:
    counter+=1
    axes[counter].hist(distrib,bins=np.arange(np.floor(min(distrib)),np.ceil(max(distrib)),1),align='mid')
    axes[counter].set_title('n = %s' %samp)
    axes[counter].yaxis.grid()
    axes[counter].set_ylabel('Frequency')
```
- Interpretation of the CLT: take a look on these 2 articles (the most clear explanations I've found). You can replicate in Python some of the examples they include.
https://towardsdatascience.com/understanding-the-central-limit-theorem-642473c63ad8.
https://www.statisticshowto.datasciencecentral.com/probability-and-statistics/normal-distributions/central-limit-theorem-definition-examples/
-Amazing plots!! Love them!! 
```Python
def ploting_crazy_binomial_PMF(k,binomial,num,k_max):
    ax[counter].plot(k,binomial,'b')
    ax[counter].grid()
    ax[counter].set_ylabel('Probability to be clicked')
    ax[counter].set_xlabel('Number of clicks')
    ax[counter].set_title('Number of users: %s' %num)
    ax[counter].fill(k,binomial,color='cyan',label='Probability Mass Function')
    ax[counter].axvline(x=k_max,color='red',label='max for %s clicks' %k_max[0])
    ax[counter].legend(ncol=2)

# Your code here
# Your code here
N=[50,500,5000]
n, p = 5, 0.15
counter=-1
fig, ax = plt.subplots(3, 1,figsize=(20,6*len(N)))
for num in N:
    counter+=1
    k=np.arange(0,num+1)
    binomial=stats.binom.pmf(k, num, p)
    max_ind=np.where(binomial==max(binomial))
    k_max=k[max_ind]
    ploting_crazy_binomial_PMF(k,binomial,num,k_max)
```
- Be careful with this: "What did you notice from the distribution plots? Comment your findings.
 Width of the curve raises when we have more samples but not by a linear relation." This is not quite correct. The conclusion on the 3 plots deals with the following: 
Normal distribution can be used as an approximation of Binomial and Poisson distributions under certain circumstances. Generally speaking:
Binomial: X ~ B(n, p) and if n is large and/or p is close to ½, then X is approximately N(np, npq)
Poisson: X ~ Po(l) then for large values of l, X ~ N(l, l) approximately.
However, we must be careful because binomial and poisson distributions are discrete while Normal distribution is continuous. We must use a continuity correction.
In this case as larger the sample size , a binomial feels more like normal distribution. Idem for poisson distributions. 
- Good one!!  Trying 2 ways of doing it!! 
```Python
mean=300 # Mean visits per day
s_obj=320  # Study objective 
num_elements=100000
# First way to do it
poss=stats.poisson.rvs(300,size=num_elements)
def find_num(x,b,c):
    """
    returns the probability of finding an element (b) in an array (x) for  a defined number of elements
    """
    return len(x[x==b])/c
print('')
print('Creating an array via poisson formulae and asking for 320: ',np.mean([find_num(poss,s_obj,num_elements) for i in range(500)]))
# Second way with pmf
print('')
print('Probability mass function for %s:                         ' %s_obj,stats.poisson.pmf(s_obj,mean))
```
- Take a look on the plot for poisson, it looks like a normal distribution, don't you think so?